### PR TITLE
update for missing a couple possibly unsafe xml parser

### LINKED ISSUE #901: Security Vulnerability - Action Required: XXE vulnerability in the newest version of the NullAway
I think your project may be vulnerable to Improper Restriction of XML External Entity Reference. It shares similarities to a recent CVE disclosure CVE-2021-3878 in the stanfordnlp/CoreNLP. The vulnerable methods are as follows:
1. com.uber.nullaway.fixserialization.XMLUtil.writeInXMLFormat(FixSerializationConfig config, String path) in the file nullaway/src/main/java/com/uber/nullaway/fixserialization/XMLUtil.java.
2. com.uber.nullaway.fixserialization.FixSerializationConfig.FixSerializationConfig(String configFilePath, int serializationVersion) in the file 
nullaway/src/main/java/com/uber/nullaway/fixserialization/FixSerializationConfig.java

**The source vulnerability information is as follows:**
> **Vulnerability Detail:**
> **CVE Identifier**: CVE-2021-3878
> **Description**: corenlp is vulnerable to Improper Restriction of XML External Entity Reference
> **Reference**:https://nvd.nist.gov/vuln/detail/CVE-2021-3878.
> **Patch:** https://github.com/stanfordnlp/corenlp/commit/e5bbe135a02a74b952396751ed3015e8b8252e99.

**Vulnerability Description:**
This vulnerability occurs because of the Improper Restriction of XML External Entity Reference. Given that the XML schema files which is compromised by a hacker, the victim conducts regular process may result in an XML External Entity (XXE) Injection attack.  

**Recommended Actions:**
The corresponding fixes are similar to CVE-2021-3878 to some extent. I have provided the following fixes by applying several patching statements, ensuring that the external entities and DTDs are not loaded when parsing and processing XML documents using the document builder. You can call the function `safeDocumentBuilderFactory` I defined below instead of directly calling `DocumentBuilderFactory.newInstance()` to create a DocumentBuilderFactory object to avoid XXE attacks.
```
  public static DocumentBuilderFactory safeDocumentBuilderFactory() {
    DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();
    try {
      dbf.setFeature("http://apache.org/xml/features/disallow-doctype-decl", true);
      dbf.setFeature("http://apache.org/xml/features/nonvalidating/load-external-dtd", false);
      dbf.setFeature("http://xml.org/sax/features/external-general-entities", false);
      dbf.setFeature("http://xml.org/sax/features/external-parameter-entities", false);
      dbf.setFeature("http://apache.org/xml/features/dom/create-entity-ref-nodes", false);
      dbf.setFeature(XMLConstants.FEATURE_SECURE_PROCESSING, true);
    } catch (ParserConfigurationException e) {
      log.warn(e);
    }
    return dbf;
  }
```

Considering the potential riskes it may have, I am willing to cooperate with your to verify, address, and report the identified vulnerability promptly through responsible means. If you require any further information or assistance, please do not hesitate to reach out to me.
Thank you and looking forward to hearing from you soon.

### PR DESCRIPTION
**Pull request by Crispy-fried-chicken**
Fixes #901

### CODE REVIEW

File: nullaway/src/main/java/com/uber/nullaway/fixserialization/FixSerializationConfig.java [hunk 1]
```diff
@@ -89,6 +89,21 @@ public FixSerializationConfig(
...
+    try {
+      dbf.setFeature("http://apache.org/xml/features/disallow-doctype-decl", true);
+      dbf.setFeature("http://apache.org/xml/features/nonvalidating/load-external-dtd", false);
+      dbf.setFeature("http://xml.org/sax/features/external-general-entities", false);
+      dbf.setFeature("http://xml.org/sax/features/external-parameter-entities", false);
+      dbf.setFeature("http://apache.org/xml/features/dom/create-entity-ref-nodes", false);
+      dbf.setFeature(XMLConstants.FEATURE_SECURE_PROCESSING, true);
+    } catch (ParserConfigurationException e) {
+      log.warn(e);
+    }
```
Comment (auke-) [inline nullaway/src/main/java/com/uber/nullaway/fixserialization/FixSerializationConfig.java:103]: Note that catching and swallowing the exception makes it possible to return a document builder factory which is still vulnerable!

File: nullaway/src/main/java/com/uber/nullaway/fixserialization/XMLUtil.java [hunk 2]
```diff
@@ -94,14 +94,25 @@ public static <T> DefaultXMLValueProvider<T> getValueFromTag(
     return new DefaultXMLValueProvider<>(null, klass);
   }
 
+  public DocumentBuilderFactory safeDocumentBuilderFactory() {
```
Comment (msridhar) [inline nullaway/src/main/java/com/uber/nullaway/fixserialization/XMLUtil.java:97]: This method should be `static`.  Also can we add a bit of Javadoc as to its purpose?

File: nullaway/src/main/java/com/uber/nullaway/fixserialization/FixSerializationConfig.java [hunk 3]
```diff
@@ -89,6 +89,17 @@ public FixSerializationConfig(
...
+    DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();
+    dbf.setFeature("http://apache.org/xml/features/disallow-doctype-decl", true);
+    dbf.setFeature("http://apache.org/xml/features/nonvalidating/load-external-dtd", false);
+    dbf.setFeature("http://xml.org/sax/features/external-general-entities", false);
+    dbf.setFeature("http://xml.org/sax/features/external-parameter-entities", false);
+    dbf.setFeature("http://apache.org/xml/features/dom/create-entity-ref-nodes", false);
+    dbf.setFeature(XMLConstants.FEATURE_SECURE_PROCESSING, true);
+    return dbf;
+  }
+
```
Comment (msridhar) [inline nullaway/src/main/java/com/uber/nullaway/fixserialization/FixSerializationConfig.java:102]: Once we make the `XMLUtil#safeDocumentBuilderFactory` method static, we can get rid of this method, and just call `XMLUtil.safeDocumentBuilderFactory()` below.

**Review by msridhar — COMMENTED** [review]
Thanks for this contribution!  I have a couple of comments that should be easy to address.  Also you will need to agree to the CLA before we can merge this contribution (see the comment on the PR by the bot).

File: nullaway/src/main/java/com/uber/nullaway/fixserialization/XMLUtil.java [hunk 4]
```diff
@@ -16,7 +16,7 @@
  * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
- * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * OUT OF OR IN CONNECTION WITH com.uber.nullaway.fixserializationTHE SOFTWARE OR THE USE OR OTHER DEALINGS IN
```
Comment (yuxincs) [inline nullaway/src/main/java/com/uber/nullaway/fixserialization/XMLUtil.java:19]: Can we revert this change? we shouldn't modify license headers unless there's a reason

File: nullaway/src/main/java/com/uber/nullaway/fixserialization/FixSerializationConfig.java [hunk 5]
```diff
@@ -89,6 +89,8 @@ public FixSerializationConfig(
     serializer = new Serializer(this, initializeAdapter(SerializationAdapter.LATEST_VERSION));
   }
 
+
+
```
Comment (yuxincs) [inline nullaway/src/main/java/com/uber/nullaway/fixserialization/FixSerializationConfig.java:93]: Can you run our configured formatter (`./gradlew spotlessApply`) on this PR?

We try to be consistent in formatting and these two blank lines don't seem right.

File: nullaway/src/main/java/com/uber/nullaway/fixserialization/XMLUtil.java [hunk 6]
```diff
@@ -94,14 +94,25 @@ public static <T> DefaultXMLValueProvider<T> getValueFromTag(
     return new DefaultXMLValueProvider<>(null, klass);
   }
 
+  public static DocumentBuilderFactory safeDocumentBuilderFactory() {
```
[the remark this answers is not in the document]
Comment (msridhar) [reply nullaway/src/main/java/com/uber/nullaway/fixserialization/XMLUtil.java:97]: JDK 21 is required in order to run certain tests.  Sorry for the hassle.

**Review by msridhar — APPROVED** [review]
Thanks for the contribution!

**Comment by msridhar** [comment]
Sorry no, I do not believe this deserves a CVE. See my previous comments.

**Pull request merged**